### PR TITLE
2085/adds top level catchAll exception filter

### PR DESCRIPTION
**Pull request by YazeedLoonat**
## Issue Overview

This PR addresses #issue
https://github.com/bloom-housing/bloom/issues/2085

- [x] This change addresses the issue in full

## Description
This adds a top level exception filter where exceptions will bubble up to/through

This is to give a console.error when errors do happen so backend errors that happen throughout the various nextjs built in logic and decorators can be easier to track down

we can add more onto here later as this is a starting point. Next steps from here would be to go through and remove try-catch blocks throughout the backend so that this top level error handler can be fleshed out/function more in-line with nextjs best practice.

I rolled logic into here from the entity not found exception filter, but in a second pass we should remove the exception filter specific to it and possibly remove this custom logic all together if its not needed. 

Also the way we have our logger attached is in line with best practice so no changes will be needed there

## How Can This Be Tested/Reviewed?
In order to review this you can run the tests to verify that all tests passed locally

then to see it in action modify one of the backend e2e tests in a way that will error (such as modifying what should be a string into a number or a similar change)

run the test and see that the console.error fires letting you know what the error is and the stack trace for it 

outside of that this shouldn't affect any current logic/behavior so without triggering something we shouldn't see any real difference

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [x] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

File: backend/core/src/filters/catch-all.filter.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { Catch, ArgumentsHost } from "@nestjs/common"
+import { BaseExceptionFilter } from "@nestjs/core"
+
+@Catch()
+export class CatchAllFilter extends BaseExceptionFilter {
+  catch(exception: any, host: ArgumentsHost) {
+    console.error({ message: exception?.response?.message, stack: exception.stack, exception })
+    if (exception.name === "EntityNotFound") {
```
Comment (pbn4) [inline backend/core/src/filters/catch-all.filter.ts:8]: I think it will never be triggered because as per NestJS docs:

> The @Catch(HttpException) decorator binds the required metadata to the exception filter, telling Nest that this particular filter is looking for exceptions of type HttpException and nothing else. The @Catch() decorator may take a single parameter, or a comma-separated list. This lets you set up the filter for several types of exceptions at once.

and 

```
@Catch(EntityNotFoundError)
export class EntityNotFoundExceptionFilter implements ExceptionFilter {
  catch(exception: EntityNotFoundError, host: ArgumentsHost) {
    const response = host.switchToHttp().getResponse()
    response.status(404).json({ message: exception.message })
  }
}
```

returns a response and by that short circuits filters chain, so CatchAllFilter will never be called when EntityNotFoundError is thrown. 

Could you rephrase that part: 

>I rolled logic into here from the entity not found exception filter, but in a second pass we should remove the exception filter specific to it and possibly remove this custom logic all together if its not needed.

What do you want to do with `EntityNotFoundExceptionFilter`?

Comment (YazeedLoonat) [reply backend/core/src/filters/catch-all.filter.ts:8]: Hey @pbn4 

for `I think it will never be triggered because as per NestJS docs:` peep https://docs.nestjs.com/exception-filters#catch-everything

when you don't pass an exception list in it catches everything

for `so CatchAllFilter will never be called when EntityNotFoundError is thrown`
I thought so to, but weirdly there was an e2e test failing where the EntityNotFoundError was only getting caught in the new catch all filter 

also for `Could you rephrase that part:` sorry, I meant that with the catch all filter in place we no longer need that specific filter anymore so we can get rid of the specific filter for it code wise. I thought maybe in the future we could get rid of that entirely and let nest's default exception list cover it for us. 

Lemme know if you have any other questions or anything like that

Comment (pbn4) [reply backend/core/src/filters/catch-all.filter.ts:8]: > for so CatchAllFilter will never be called when EntityNotFoundError is thrown
I thought so to, but weirdly there was an e2e test failing where the EntityNotFoundError was only getting caught in the new catch all filter

OK I see what the problem was, the `EntityNotFoundExceptionFilter` is never called because TypeORM does not throw a proper EntityNotFoundError class instance, it throws a JS object. A proper implementation of this filter should be 
```
@Catch(NotFoundException)
export class NotFoundExceptionFilter implements ExceptionFilter {
  catch(exception: NotFoundException, host: ArgumentsHost) {
    const response = host.switchToHttp().getResponse()
    response.status(404).json({ message: exception.message })
  }
}
```

because most of the services use findOne and check whether returned object is undefined and then throw NotFoundException. 

Anyway your approach is better so let's keep that, you can remove `EntityNotFoundExceptionFilter` completely in this PR.

**Review by pbn4 — APPROVED** [review]

**Review by seanmalbert — APPROVED** [review]
